---
title: "Mitocall Pipeline"
author: "Jip van Dinter"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r, include = FALSE}
setwd('E:/mito_internship/')
load("mito_project.RData")
```

## Introduction

This vignette serves as a tutorial for a pipeline to call mutations in mitochondrial DNA adapted from the [GATK pipeline](https://github.com/gatk-workflows/gatk4-mitochondria-pipeline). It will guide you how to process your __WGS hg38 BAM file__ into an __R dataframe__ containing potential somatic mutations. In addition it will show some basic plotting functions to visualise the data and some models I have created based on my own data. 

__Required input__:

 - hg38 WGS BAM file
 - File with median / mean coverage of the autosome from the IAP pipeline
 - Reference fasta 
 - List / table of samples and additional data 

__Required software__:

- __For the WDL pipeline__:
  + Cromwell v29 or later
  + GATK 4.1.2 or later
  + SGE backend (SLURM backend support might be released)
- __For the analysis__:
  + Latest R version
  + The analysis in R requires the following packages:
  
```{r eval = FALSE} 
library(tidyverse)
library(ggbeeswarm)
library(MutationalPatterns)
library(VariantAnnotation)
library(ggpubr)
library(ggrepel)
library(ggsci)
library(RColorBrewer)
library(lme4)
library(nlme)
library(regioneR)
library(karyoploteR)
library(magrittr)
library(car)
library(BSgenome.Hsapiens.UCSC.hg38)
library(gridExtra)
```

## Setting up the pipeline folder structure

Before we can run the pipeline, we first need to set up our base of operations; the workflow directory. Download all the files required from my [GitHub page](https://github.com/jvandinter/MitoCallPipeline) and put them in your chosen directory like this:

 __MAIN directory__
 
- `application.conf.mito.31`
- `mitopipeline.config`
- `mitolib-0.1.2.jar`
- `mito_pipeline_start.sh`
- `mito_pipeline_run.sh`
- `final_mito_output.sh`
- `WDL folder`
- `INPUT folder`
 
 __WDL directory__
 
- `AlignAndCall.wdl`
- `MitochondriaPipeline.wdl`
- `AlignmentPipeline.wdl`

__Inputs directory__

- `finish_mitocall_jsons.py`
- `mitochondrial_workflow_inputs_unfinished.json`
- (The `options.json` will be generated per sample)

__Resources directory__

- `Blacklist files provided` [here](https://github.com/jvandinter/MitoCallPipeline/bucket/)
- `MT pipeline resource files from the` [GATK bucket](https://github.com/jvandinter/MitoCallPipeline/bucket/) `I have hosted`
- `This does not have to be in the` ___MAIN directory___`, it can go anywhere else as long as you update mitochondrial_workflow_inputs_unfinished.json`

>__!!!__ If you do not have the mitochondrial resources directory set up yet, open `mitochondrial_workflow_inputs_unfinished.json` in your favorite text editor and change the location of all files to your own chosen location to unpack the required sequence files.

---

Open the `mitopipeline.config` file and edit the path of the following variables:

- `MAIN_LOC`
  + This is the main directory which contains all the required files for the analysis
- `WORK_LOC`
  + This is the directory name which will contain all the output folders
- `INPUT_LOC`
  + The inputs and options JSON directory. Also should include the python script and the unfished input JSON.
- `WDL_LOC`
  + The directory name where all the pipeline WDLs are stored
- `CROMWELL_LOC`
  + The location of the cromwell jar file on the HPC.
- `HG38_MAP_LOC`
  + This is the folder which contains the hg38 IAP output and where the scripts searches for the WGSMetrics.txt file. If it is not present, please re-generate the summary statistics files for optimal variant calling, otherwise the coverage will be set to a default 30.
- `EMAIL`
  + This is the email that will be used in the annotate_vcf steps
  
---

### Run the script

You can now run the pipeline with the following command on the HPC:

```{r, eval = FALSE}
cd `Insert your MAIN_LOC here`
```

```{r, eval = FALSE}
qsub mito_pipeline_start.sh `your email`
```

Each sample will take on average 30 minutes

> __!!!__ If the files fail due to abort and time constrictions, just submit the job again. The script will automatically detect which samples have been finished and continue on with the others.

---

When they are all finished, you have to retrieve the output from all subfolders by running the following command:

```{r, eval = FALSE}
qsub final_mito_output.sh
```

The script will automatically get the output directory from the `mitopipeline.config`.

---

### Optional: annotate VCF files using the vanBoxtel IAP

First, run the following command: 

```{r, eval = FALSE}
qsub convert_chrM_to_MT.sh`
```

This script will do two things:

1) Change chrM to MT in all VCFs

2) Annotate the variants located in the VCFs

All the annotated VCF files are located in `MAIN_LOC/annotated_vcfs/`. 

---

## Downstream analysis

Firstly, copy the output and annotated_vcfs files from the HPC to your local computer or source them from the HPC mount.

__Required files:__

- All (annotated) VCF files
- WGSMetrics text file per sample (if interested in copy number variation)
- All function files found on [GitHub](https://github.com/jvandinter/MitoCallPipeline/bucket/)

```{r include = FALSE} 
library(tidyverse)
library(ggbeeswarm)
library(MutationalPatterns)
library(VariantAnnotation)
library(ggpubr)
library(ggrepel)
library(ggsci)
library(RColorBrewer)
library(lme4)
library(nlme)
library(regioneR)
library(karyoploteR)
library(magrittr)
library(car)
library(BSgenome.Hsapiens.UCSC.hg38)
library(gridExtra)
```

When building your own sample reference data frame, follow the structure of the following DF, otherwise the functions in this vignette will not work:

```{r echo=FALSE} 
head(sample_reference_df[,1:5])
```

Initiate the following parameters in R:

```{r eval=FALSE}
plotdir <- 'E:/mito_internship/plots'
vcf_wd <- 'E:/mito_internship/_annotated_vcfs'
wgs_wd <- 'E:/mito_internship/output'
function_wd <- 'E:/mito_internship/scripts/Restructured/Downstream_post_processing/functions'
vcf_filter <- "*\\_snpSift.vcf$" # To select annotated VCFs
ref_genome <- "BSgenome.Hsapiens.UCSC.hg38"
low_vaf_filter <- 0.01 # The lower and upper bound of VAF filtering, as by GATK recommendation
high_vaf_filter <- 0.99
vcf_filter <- '.*\\.vcf$' # or '.*\\snpSift$' when using the annotated files
max_missing_calls <- 3  # Threshold of calls when to look for the variant in unfiltered calls
```

When all parameters are set up, you can execute the following commands to load all functions from your function directory:

```{r eval=FALSE} 
setwd(function_wd)
files_list <- list.files(pattern = '.R$', ignore.case = T)
sapply(files_list, source, .GlobalEnv)
```

Let's first take a look at the coverage of our samples, to see whether they reach the threshold of at least 1000 coverage:

```{r eval=FALSE}
cnv_df <- calculate_mito_cnv(wd = wgs_wd, 
                             mito_reference_df = sample_reference_df)
```

To plot the coverage, you can use the following code:

```{r, fig.width= 9, fig.height = 7, fig.align= 'center'}
plot_coverage = ggplot(cnv_df, 
                       aes(x = patient, 
                           y = mt_mean)) +
  geom_boxplot(outlier.shape = NA) +
  geom_quasirandom(aes(col = sample, shape = state), 
                   size = 3) +
  geom_hline(yintercept = 1000, 
             col = 'red', 
             linetype = 2) +
  xlab("Patient Source") +
  ylab("Mean MT Coverage (log10)") +
  theme_BM() +
  guides(color=FALSE) +
  scale_y_log10(labels = scales::comma)

plot_coverage
```

The following code snippet can help you to save the plots generated:

```{r eval=FALSE}
plotname <- paste(plotdir,'optional_folder_name','name_of_plot', sep ='/')

# Put your plot code here and save it as a variable
plot = ggplot(data, aes( x = x, y = y)) +
  geom_point()

ggsave(paste(plotname, 
             'pdf', 
             sep = '.'), 
       width = 11.69,
       height = 8.27,
       plot, 
       device = 'pdf')
```

In my experience, samples that had less than 1000 coverage either outputted almost no somatic variants at all, or had many false positives. Best to remove the samples that did not meet the coverage threshold from your sample reference data frame with the following command:

```{r eval=FALSE}
removed_samples <- cnv_df[,cnv_df$cnv_mean < 1000]$sample
sample_reference_df <- filter(sample_reference_df, !(sample %in% removed_samples))
```

With the final sample list, we can start loading 

---

### Loading the data

Use the following functions to load the VCF variants into R:

```{r eval=FALSE}
passed_vcflist <- filter_passed_variants(ref_genome = ref_genome, 
                                   wd = vcf_wd, 
                                   vcf_filter = vcf_filter)

all_variants <- grab_all_variants(ref_genome = ref_genome,
                             wd = vcf_wd,
                             vcf_filter = vcf_filter,
                             mito_reference_df = sample_reference_df)

annotated_variants <- grab_variant_annotations(vcf_filter = vcf_filter)
```

These functions will usew the files for various things:

- __Per_base_coverage__
  + Output is a data frame containing a base-per-base coverage overview for each sample
- __Get_vaf_df__
  + Output is a data frame which holds the Variant Allele Frequency per variant
- __create_variant_df__
  + Take all PASSED variants from passed_vcflist and puts them in a data frame. Annotates variants with sample of origin, patient origin, etc.
- __filtered_variants__
  + Filter variants based on criteria named in [the report](github.com/MitoCallPipeline/pdf/here)
- __somatic_variant_df__
  + Filter variants on VAF

```{r eval=FALSE}
# For QC coverage plots
per_base_coverage <- calculate_mt_coverage(wd = wgs_wd, 
                                  mito_reference_df = sample_reference_df)
# Calculate VAF of each variant
vaf_df <- get_vaf_df(vcf_list = passed_vcflist)

# Load passed variants in DF
variant_df <- create_variant_df(vcf_list = passed_vcflist, 
                                mito_reference_df = sample_reference_df)

# Filter variants (see final report)
filtered_variants <- filter_variants(variant_df = variant_df,
                              vcf_all_df = all_variants,
                              sample_reference_df = sample_reference_df)

# Filter variants on VAF
somatic_variant_df <- combine_mito_data(shared_mutation_output = filtered_variants,
                         vaf_output = vaf_df,
                         low_vaf_filter = low_vaf_filter,
                         high_vaf_filter = high_vaf_filter)

indel_df <- filter(somatic_variant_df, 
                   nchar(somatic_variant_df$ref) > 1 | nchar(somatic_variant_df$alt) > 1)
indel_df$indel <- ifelse(nchar(indel_df$ref) > nchar(indel_df$alt), 
                         'deletion', 
                         'insertion')
indel_annotate <- left_join(indel_df, annotated_variants,
                               by = c('variant', 'sample'))

# Remove INDELs from somatic_variant_df

somatic_variant_df <- anti_join(somatic_variant_df, indel_df, 
                 by = 'variant')
somatic_variant_annotate <- left_join(somatic_variant_df, annotated_variants, 
                          by = c('variant', 'sample'))

# Calculate number of mutations per sample

total_sample_mutation_freq <- count_variant_occurences(combined_mito_df = somatic_variant_df,
                                                       mito_reference_df =sample_reference_df)
```

This will create a coverage plot for each sample:

```{r eval=FALSE}
sapply(sample_reference_df$sample, function(x) { 
  plot_cov(per_base_coverage, 
           x, 
           paste(plotdir,'Coverage',sep = '/'))
})
```

Like so:

```{r, fig.width= 9, fig.height = 7, fig.align= 'center'}
plot_cov(per_base_coverage, 'N01', paste(plotdir,'Coverage',sep = '/'))
```

Here, we split the INDELs from the SNVs and annotate the variants where applicable.

- ___count_variant_occurences___
  + Calculates the number of somatic mutations per sample

```{r eval=FALSE}
# Split SNPs and INDELs

indel_df <- filter(somatic_variant_df, 
                   nchar(somatic_variant_df$ref) > 1 | nchar(somatic_variant_df$alt) > 1)
indel_df$indel <- ifelse(nchar(indel_df$ref) > nchar(indel_df$alt), 
                         'deletion', 
                         'insertion')
indel_annotate <- left_join(indel_df, annotated_variants,
                               by = c('variant', 'sample'))

# Remove INDELs from somatic_variant_df

somatic_variant_df <- anti_join(somatic_variant_df, indel_df, 
                 by = 'variant')
somatic_variant_annotate <- left_join(somatic_variant_df, annotated_variants, 
                          by = c('variant', 'sample'))

# Calculate number of mutations per sample

total_sample_mutation_freq <- count_variant_occurences(combined_mito_df = somatic_variant_df,
                                                       mito_reference_df =sample_reference_df)
```

---

To plot an overview of the number of mutations per group, you can use:

```{r, fig.width= 9, fig.height = 7, fig.align= 'center'}
plot <- ggplot(subset(total_sample_mutation_freq, 
                      state %in% c('healthy','trisomy','AML'))) +
  geom_histogram(aes(x = freq), 
                 fill = 'orange', 
                 binwidth = 1) +
  theme_BM(base_size = 30) +
  xlab('Mutations per sample') +
  ylab('Sample count') +
  facet_wrap(. ~ state)

plot
```

To make work with the data easier, I create new dataframes for specific analyses. To keep this vignette succinct, I will only give 1 single example per analysis.

```{r eval=FALSE}
# Subset DFs to only healthy variants
healthy_variants <- subset(somatic_variant_annotate, 
                           state == 'healthy' &
                           !(patient == 'AHH1') &
                           bulk == 'clone')
healthy_cnv <- subset(cnv_df, 
                      state == 'healthy' &
                      !(patient == 'AHH1') &
                      mt_mean > 1000 &
                      bulk == 'clone')
healthy_freq <- subset(total_sample_mutation_freq, 
                       state == 'healthy' &
                       !(patient == 'AHH1') &
                       bulk == 'clone')

# Subset DFs to only foetal variants
foetus_cnv <- subset(cnv_df, 
                     state %in% c('healthy','trisomy') &
                     patient %in% c('N01','OS1','MH2','MH3','NR1','NR2') &
                     !(sample %in% foetus_remove) &
                     mt_mean > 1000 &
                     bulk == 'clone')
foetus_freq <- subset(total_sample_mutation_freq, 
                      state %in% c('healthy','trisomy') &
                      patient %in% c('N01','OS1','MH2','MH3','NR1','NR2') &
                      bulk == 'clone')
```

---

## The Analyses

### Mutational spectra

To plot the mutation spectra with ___MutationalPatterns___ and ___KaryoplotR___, the variants have to be in a Granges object. A subset of the data is made to look for possible strand bias. The Granges objects will be generated with the following code:

```{r eval=FALSE}
healthy_c_t <- subset(healthy_variants, ref %in% c('C','T'))

healthy_g_a <- subset(healthy_variants, ref %in% c('G','A'))

healthy_granges <- unique(GRanges(seqnames = Rle("chrM", nrow(healthy_variants)),
                    ranges = healthy_variants$mut_pos,
                    strand = rep('*', nrow(healthy_variants)),
                    ref = healthy_variants$ref,
                    alt = healthy_variants$alt))

healthy_ct_granges <- unique(GRanges(seqnames = Rle("chrM", nrow(healthy_c_t)),
                      ranges = healthy_c_t$mut_pos,
                      strand = rep('*', nrow(healthy_c_t)),
                      ref = healthy_c_t$ref,
                      alt = healthy_c_t$alt))

healthy_ga_granges <- unique(GRanges(seqnames = Rle("chrM", nrow(healthy_g_a)),
                      ranges = healthy_g_a$mut_pos,
                      strand = rep('*', nrow(healthy_g_a)),
                      ref = healthy_g_a$ref,
                      alt = healthy_g_a$alt))
```

___single_type_occurences___ will output a count dataframe with count data for each type of mutation.

```{r eval=FALSE}
healthy_mutprof <- single_type_occurence(healthy_granges,
                                          ref_genome,
                                          healthy_freq)
healthy_ct_mutprof <- single_type_occurence(healthy_ct_granges,
                                    ref_genome,
                                    healthy_freq)
healthy_ga_mutprof <- single_type_occurence(healthy_ga_granges,
                                    ref_genome,
                                    healthy_freq)

healthy_combined <- rbind(healthy_mutprof, healthy_ct_mutprof, healthy_ga_mutprof)

healthy_combined <- rbind(healthy_mutprof, healthy_ct_mutprof, healthy_ga_mutprof)
```

Finally, plot the mutational spectra with the following code:

```{r, fig.width= 9, fig.height = 7, fig.align= 'center'}
plot <- plot_spectrum(healthy_combined, CT = TRUE, by = c('healthy','ct_healthy','ga_healthy'))
plot
```

---

### Rainfall plot

Rainfall plots show the genomic distribution of mutations and visualises clusters of mutations close together. By replacing __healthy_grangres__ with __healthy_ct_granges__ or your equivalent, you can see whether the distribution is different for G/A or C/T reference nucleotides.

```{r, fig.width= 9, fig.height = 7, fig.align= 'center'}
COLORS6 = c('C>A' = "#2EBAED", 'C>G' = "#000000", 'C>T' = "#DE1C14",'T>A' = '#D4D2D2',
           'T>C' = "#ADCC54", 'T>G' = "#F0D0CE")

healthy.colors <- getVariantsColors(healthy_granges$ref, healthy_granges$alt, color.table = COLORS6)

kp <- plotKaryotype(chromosomes = 'chrM', 
                    genome = 'hg38', 
                    ideogram.plotter = NULL,
                    labels.plotter = NULL)
kpAddCytobandsAsLine(kp)
kpAddChromosomeNames(kp, 
                     srt=45)
kpAxis(kp, ymax = 7, 
       tick.pos = 1:7, 
       cex = 2)
kpPlotRainfall(kp, 
               data = healthy_granges, 
               col = healthy.colors, 
               cex = 1.2)
kpAddLabels(kp, labels = c("Distance between mutations (log10)"), 
            srt=90, 
            pos=1, 
            label.margin = 0.06, 
            cex = 1.5)

```

---

### GLMM Modelling

A generalised linear mixed effects model (GLMM) is used to see here whether age has an effect on mutation frequency. A GLMM is used, due to the fact that it can handle the sparse count data (poisson distribution). A linear mixed effects model is used as it handles the patients with different sample numbers correctly. Here, mutation frequency (freq) is the response variable, age is the random variable, and patient is the grouping variable. A random slope based on age is calculated.

```{r eval=FALSE}
ageline_freq_model <- glmer(freq ~ age + (0 + age | patient), 
                      data = healthy_freq, 
                      family = poisson)
```

By plotting the model residuals we can check whether there is any bias in the sample. In general, a random distribution over the plot indicates no bias in the dataset and model.

```{r, fig.width= 9, fig.height = 7, fig.align= 'center'}
plot(ageline_freq_model, id = 0.05, idLabels = healthy_freq$sample)
```

This code snippet will convert the residuals to Z scores and outputs the two-sided P value that is associated with that particular P value to check for significant outliers.

```{r eval = FALSE}
healthy_freq$resids <- resid(healthy_freq_model, type = "pearson") %>% 
  abs() %>% 
  multiply_by(-1) %>% 
  pnorm() %>% 
  multiply_by(2) %>% 
  p.adjust(method = "fdr")
```

Save the summary in a variable and add the fitted values to the dataframe for plotting.

```{r}
sum_ageline_freq <- summary(ageline_freq_model)

healthy_freq$fit <- predict(ageline_freq_model, re.form=NA) 

pval <- formatC(sum_ageline_freq$coefficients[2,4],
               digits = 6,
               format = 'f')
```

Finally, plot the model:

```{r, fig.width= 9, fig.height = 7, fig.align= 'center'}
# Model healthy variant ageline plot

plotname <- paste(plotdir,'model','MT_model_healthy_ageline', sep ='/')

plot_ageline <- ggplot(healthy_freq,
                    aes(x = age, y = freq )) +
  geom_jitter(aes(col = patient), 
              size = 4, 
              width = 1.5, 
              height = 0.3) +
  geom_hline(yintercept = c(-0.5,0.5,1.5,2.5), 
             linetype = 2, 
             alpha = 0.5) +
  geom_line(aes(y = fit), 
            size = 1.5, 
            color = 'red') +
  ylab("SNVs per genome") +
  xlab("Age (years)") +
  guides(color = F) +
  annotate(geom='text', 
           x = 6, 
           y = 3.5,
           label = paste('P',pval, sep = ' = '), 
           size = 5) +
  theme_BM()

plot_ageline

ggsave(paste(plotname, 
             'pdf', 
             sep = '.'), 
       width = 11.69,
       height = 8.27,
       plot_ageline, 
       device = 'pdf')
```

---

## 

Here is a different GLMM model created. In addition to age (as it was significant in the previous model), we want to look whether the disease state has any effect on the number of somatic mutations:

```{r eval = FALSE}
foetus_freq$state <- factor(foetus_freq$state, levels = c('healthy','trisomy'))

foetus_freq_model <- glmer(freq ~ state + age + (0 + age | patient), 
                           data = foetus_freq, 
                           family = poisson)
```

Plot the residuals, annotating any possible outliers:

```{r, fig.width= 9, fig.height = 7, fig.align= 'center'}
plot(foetus_freq_model, id = 0.05, idLabels = foetus_freq$sample)

```

Find significant outliers and save model data in variables.

```{r}
foetus_freq$resids <- resid(foetus_freq_model, type = "pearson") %>% 
  abs() %>% 
  multiply_by(-1) %>% 
  pnorm() %>% 
  multiply_by(2) %>% 
  p.adjust(method = "fdr")

sum_foetus_freq <- summary(foetus_freq_model)

foetus_freq$fit <- predict(foetus_freq_model, re.form=NA)

pval <- formatC(sum_foetus_freq$coefficients[2,4], digits = 4, format = 'f')

```

Finally, plot the model.

```{r, fig.width= 9, fig.height = 7, fig.align= 'center'}

plot_foetus_freq <- ggplot(foetus_freq,
                     aes(x = age * 52, 
                         y = freq )) +
  geom_line(aes(y = fit, col = state), 
            size = 1.5) +
  geom_quasirandom(aes(col = state, shape = state), 
                   size = 4, 
                   alpha = 0.5, 
                   width = 0.4, 
                   groupOnX = F) +
  geom_hline(yintercept = c(-0.5,0.5,1.5), 
             linetype = 2, 
             alpha = 0.5) +
  ylab(" ") +
  xlab("Age (weeks)") +
  annotate(geom='text', 
           x= 10.5, 
           y = 2,
           label = paste('P',pval, sep = ' = '), 
           size = 5) +
  theme_BM() +
  guides(fill = F)

plot_foetus_freq
```

---

### LME Modelling

Because the copy number is a continuous variable, the model is created with regular linear-mixed effects modelling using the __NLME__ R package. The function notation is slightly different, as you have to split the random component of your model. Here, the response variable, age a random effect and patient the grouping factor. A random slope for age will be modelled. 

```{r eval=FALSE}
ageline_cnv_model <- lme(cnv_mean ~ age, 
          random = ~ 0 + age | patient, 
          data = healthy_cnv)
```

Plot the residuals, where potential outliers are annotated.

```{r, fig.width= 9, fig.height = 7, fig.align= 'center'}
plot(ageline_cnv_model, id = 0.05, idLabels = healthy_cnv$sample)
```

Add Z-score p values based on the residuals to the dataframe, and save some model data in variables.

```{r}
healthy_cnv$resids <- resid(ageline_cnv_model, type = "pearson") %>% 
  abs() %>% 
  multiply_by(-1) %>% 
  pnorm() %>% 
  multiply_by(2) %>% 
  p.adjust(method = "fdr")

sum_ageline_cnv <- summary(ageline_cnv_model)
 
healthy_cnv$fit <- predict(ageline_cnv_model, level = 0)

pval <- formatC(sum_ageline_cnv$tTable[2,5],
                digits = 3,
                format = 'f')
```

Finally, plot the model:

```{r, fig.width= 9, fig.height = 7, fig.align= 'center'}
# Model healthy cnv ageline plot
plot_healthy_cnv = ggplot(healthy_cnv,
                    aes(x = age, 
                        y = cnv_mean )) +
  geom_jitter(aes(col = patient),
              size = 4, 
              width = 1.5, 
              height = 0.3) +
  geom_line(aes(y = fit), 
            color = 'red', 
            size = 1) +
  annotate(geom='text', 
           x = 40, 
           y = 600,
           label = paste('P',pval, sep = ' = '), 
           size = 5) +
  ylab("mtDNA Copy Number") +
  xlab("Age (years)") +
  theme_BM() +
  guides(color = F)
plot_healthy_cnv
```

---

If you have any further questions, reach out to me or Freek Manders.